karapace/karapacemetrics.py

```diff
@@ -0,0 +1,169 @@
+"""
+karapace - metrics
+Supports collection of system metrics
+list of supported metrics:
+connections-active - The number of active HTTP(S) connections to server.
+                     Data collected inside aiohttp request handler.
+
+Copyright (c) 2023 Aiven Ltd
+See LICENSE for details
+"""
+from datetime import datetime
+from kafka.metrics import MetricName, Metrics
+from kafka.metrics.measurable_stat import AbstractMeasurableStat
+from kafka.metrics.stats import Avg, Max, Rate, Total
+from karapace.config import Config
+from karapace.statsd import StatsClient
+from typing import Dict, Optional
+
+import schedule
+import threading
+import time
+
+
+class Value(AbstractMeasurableStat):
+    """
+    An AbstractSampledStat that maintains a simple average over its samples.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.value = 0.0
+
+    # pylint: disable=unused-argument
+    def measure(self, config: object, now: int) -> float:
+        return self.value
+
+    def record(self, config: object, value: float, time_ms: int) -> None:
+        self.value = value
+
+
+class Singleton(type):
+    _instances: Dict["Singleton", "Singleton"] = {}
```

Comment:
```suggestion
    _instances: Dict["type[Singleton]", "Singleton"] = {}
```

issue: I don't believe we're mapping instance to instance here, but rather class to instance.

Reply:
> Continuing discussion from [original PR](https://github.com/aiven/karapace/pull/625#issuecomment-1565252743). If possible, in the future I'd ask that you please avoid closing and re-opening separate PRs like this, as it makes it quite hard to follow progress and review. If you need to "start over", please feel free to force push to your feature branch instead.
> 
> > Karapace competitors use sampled metrics and perform calculations for avg, min, max on the application side in order to facilitate Pull-based statistics services. This particular model is commonly utilized in platforms like Prometheus [...]
> 
> This does not sound accurate to me. In my experience when using Prometheus, it's much more common to avoid calculations on the application side, and expose metrics in their raw forms. Averages, and other aggregations are then produced by posing PromQL queries to the Prometheus service.
> 
> It seems reasonable to take that route here as well, and for instance rather expose `sent_response_bytes`, `received_request_bytes`. Together with a request count, that should be everything required to reliable calculate aggregations in Prometheus or similar.

My client requires support for metrics that are already supported in competitor projects. Additionally, they anticipate the option to integrate with Prometheus in the future. Consequently, I have implemented the required support accordingly. The test results using Graphite, which I used for testing purposes, show the expected data output. For Prometheus integration, we have two options: utilizing the statsd-prometheus module or developing our own converter to send aggregated data to Prometheus. I believe Prometheus has the capability to handle aggregated data, allowing us to aggregate the metrics as required.

Reply:
done

Reply:
but my way worked and passed all pypy checks

Reply:
Hmm right, that's another issue because the file does not have `from __future__ import annotations`, but I was also off in my comment.

`cls` is equal to `KarapaceMetrics` here, which is an instance of `Singleton` because that is its metaclass (`type(KarapaceMetrics) is Singleton`). So the key type is actually correct, however the value type isn't. You can verify this by checking `assert isinstance(instance, Singleton)` in `__call__`.

I'd suggest updating the return type of `__call__`, as well as the value type of `_instances` to `object`.

Reply:
For me also wors following construction:
    _instances: Dict[type("Singleton"), "Singleton"] = {}

Reply:
Construction     _instances: dict[type[object], Singleton] = {} works

Reply:
@libretto That doesn't make sense though, the key is overly broad and the value is incorrect? I was proposing `dict[Singleton, object]`.

Reply:
I need a dict where keys are exactly represented, classes... But seems Your offer also looks like dict[instance,instance]. So let me try to get rid of dict at all.

Reply:
> So let me try to get rid of dict at all.

No need to do this.

> I need a dict where keys are exactly represented, classes... But seems Your offer also looks like dict[instance,instance].

No, as stated earlier, I was off in my first comment. You can verify that `assert isinstance(cls, Singleton)` passes inside the method. (This is because `type(KarapaceMetrics) == Singleton`).